File: javascript/nucleotide-count/nucleotide-count.js
export class NucleotideCounts {
    static parse(strand) {
        let a = 0;
        let c = 0;
        let g = 0;
        let t = 0;

        for (const char of strand) {
            if (char == 'A')
                a += 1;
            else if (char == 'C')
                c += 1;
            else if (char == 'G')
                g += 1;
            else if (char == 'T')
                t += 1;
            else
                throw "Invalid nucleotide in strand";
        }

        return `${a} ${c} ${g} ${t}`;
    }
}

File: lua/nucleotide-count/nucleotide-count.lua
local DNA = {}

function DNA:new(strand)
  new = {
    nucleotideCounts = { A = 0, T = 0, C = 0, G = 0 }
  }

  for i = 1, #strand do
    local c = strand:sub(i, i)
    new.nucleotideCounts[c] = new.nucleotideCounts[c] + 1
  end

  self.__index = self
  return setmetatable(new, self)
end

function DNA:count(nucleotide)
  if not self.nucleotideCounts[nucleotide] then
    error("Invalid Nucleotide")
  end

  return self.nucleotideCounts[nucleotide]
end

return DNA
